lib/SPIRV/SPIRVLowerBitCastToNonStandardType.cpp

```diff
@@ -0,0 +1,183 @@
+        if (auto *DestVecTy = dyn_cast<VectorType>(DestTy)) {
+          uint64_t NumElemsInDestVec = DestVecTy->getElementCount().getValue();
+          if (!isValidVectorSize(NumElemsInDestVec))
+            BCastsToNonStdVec.push_back(std::make_pair(&I, DestVecTy));
+        }
+      }
+
+    for (auto &I : BCastsToNonStdVec)
+      Changed |= lowerBitCastToNonStdVec(I, InstsToErase);
+
+    for (int I = InstsToErase.size() - 1; I >= 0; I--)
+      InstsToErase[I]->eraseFromParent();
+
+    return Changed ? PreservedAnalyses::none() : PreservedAnalyses::all();
+  }
+
+  bool lowerBitCastToNonStdVec(std::pair<Instruction *, VectorType *> &Inst,
+                               std::vector<Instruction *> &InstsToErase) {
+    Instruction *I = Inst.first;
+    InstsToErase.push_back(I);
+    auto *InstIter = I;
+    IRBuilder<> Builder(I);
+    Value *NewSrc = I->getOperand(0);
+    // It is assumed that the function can contain addrspacecast after handled
+    // bitcast instruction, so addrspacecast should also be handled
+    for (auto *U : I->users()) {
+      if (auto *ASCastInst = dyn_cast<AddrSpaceCastInst>(U)) {
+        unsigned DestAddrSpace = ASCastInst->getDestAddressSpace();
+        auto *SrcTy = cast<PointerType>(NewSrc->getType())->getElementType();
+        auto *NewDestPtrTy = SrcTy->getPointerTo(DestAddrSpace);
+        NewSrc = new AddrSpaceCastInst(NewSrc, NewDestPtrTy);
+        Builder.Insert(NewSrc);
+        InstIter = ASCastInst;
+        InstsToErase.push_back(InstIter);
+        break;
+      }
+    }
+    auto *SrcTy = cast<PointerType>((NewSrc)->getType())->getElementType();
+    auto *Load = Builder.CreateLoad(SrcTy, NewSrc);
+    // In the already known pattern, the bitcast is followed by load instruction
+    for (auto *U : InstIter->users()) {
+      if (auto *LI = dyn_cast<LoadInst>(U)) {
+        InstIter = LI;
+        InstsToErase.push_back(InstIter);
+        break;
+      }
+    }
+    // The bitcast is followed by extractelement instruction
+    VectorType *DestVecTy = Inst.second;
+    uint64_t NumElemsInDestVec = DestVecTy->getElementCount().getValue();
+    uint64_t NumElemsInSrcVec =
+        cast<VectorType>(SrcTy)->getElementCount().getValue();
+    int ElemIdx = 0;
+    Instruction *ExtrElInstIter = InstIter;
+    for (auto *U : InstIter->users()) {
+      if (auto *EEI = dyn_cast<ExtractElementInst>(U)) {
+        ElemIdx = cast<ConstantInt>(EEI->getIndexOperand())->getSExtValue() /
+                  (NumElemsInDestVec / NumElemsInSrcVec);
+        ExtrElInstIter = cast<Instruction>(U);
+        InstsToErase.push_back(ExtrElInstIter);
+        break;
+      }
+    }
+    auto *ExtractElement = Builder.CreateExtractElement(
+        Load, ConstantInt::get(Type::getInt64Ty(I->getContext()), ElemIdx));
+    auto *Trunc =
+        Builder.CreateTrunc(ExtractElement, DestVecTy->getElementType());
+    ExtrElInstIter->replaceAllUsesWith(Trunc);
+    return true;
+  }
+};
```

Comment:
```suggestion
  bool lowerBitCastToNonStdVec(Instruction *Inst,
                               std::vector<Instruction *> &InstsToErase) {
    Instruction *OldInstIter{Inst};
    IRBuilder<> Builder(OldInstIter);
    Value *NewValue = OldInstIter->getOperand(0);
    VectorType *NewVecTy = getVectorType(NewValue->getType());
    InstsToErase.push_back(OldInstIter);

    // Handle addrspacecast instruction after bitcast if present
    for (auto *U : OldInstIter->users()) {
      if (auto *ASCastInst = dyn_cast<AddrSpaceCastInst>(U)) {
        unsigned DestAS = ASCastInst->getDestAddressSpace();
        auto *NewVecPtrTy = NewVecTy->getPointerTo(DestAS);
        // AddrSpaceCast is created explicitly instead of using method
        // IRBuilder<>.CreateAddrSpaceCast because IRBuilder doesn't create
        // separate instruction for constant values. Whereas SPIR-V translator
        // doesn't like several nested instructions in one.
        NewValue = new AddrSpaceCastInst(NewValue, NewVecPtrTy);
        Builder.Insert(NewValue);

        OldInstIter = ASCastInst;
        InstsToErase.push_back(OldInstIter);
        break;
      }
    }

    // Handle load instruction which is following the bitcast in the pattern
    for (auto *U : OldInstIter->users()) {
      if (auto *LI = dyn_cast<LoadInst>(U)) {
        NewValue = Builder.CreateLoad(NewVecTy, NewValue);

        OldInstIter = LI;
        InstsToErase.push_back(OldInstIter);
        break;
      }
    }

    // Handle extractelement instruction which is following the load
    for (auto *U : OldInstIter->users()) {
      if (auto *EEI = dyn_cast<ExtractElementInst>(U)) {
        VectorType *OldVecTy = getVectorType(Inst->getType());
        uint64_t NumElemsInOldVec = OldVecTy->getElementCount().getValue();
        uint64_t NumElemsInNewVec = NewVecTy->getElementCount().getValue();
        uint64_t ElemIdx =
            cast<ConstantInt>(EEI->getIndexOperand())->getSExtValue() /
                (NumElemsInOldVec / NumElemsInNewVec);
        NewValue = Builder.CreateExtractElement(NewValue, ElemIdx);
        NewValue = Builder.CreateTrunc(NewValue, OldVecTy->getElementType());

        OldInstIter = EEI;
        InstsToErase.push_back(OldInstIter);
        break;
      }
    }

    OldInstIter->replaceAllUsesWith(NewValue);
    return true;
  }
};
```

Reply:
I also think that this function may be not a method because there are no dependencies on SPIRVLowerBitCastToNonStandardTypePass class.